src/openssl/cipher.cr

```diff
@@ -0,0 +1,125 @@
+require "./openssl"
+
+class OpenSSL::Cipher
+  class Error < OpenSSL::OpenSSLError
+  end
+
+  def initialize(name)
+    cipher = LibCrypto.evp_get_cipherbyname name
+    raise ArgumentError.new "unsupported cipher algorithm #{name.inspect}" unless cipher
+
+    @ctx = LibCrypto.evp_cipher_ctx_new
+    # The EVP which has EVP_CIPH_RAND_KEY flag (such as DES3) allows
+    # uninitialized key, but other EVPs (such as AES) does not allow it.
+    # Calling EVP_CipherUpdate() without initializing key causes SEGV so
+    # we set the data filled with "\0" as the key by default.
+    cipherinit cipher: cipher, key: "\0" * LibCrypto::EVP_MAX_KEY_LENGTH
+  end
+
+  def encrypt
+    cipherinit enc: 1
+  end
+
+  def decrypt
+    cipherinit enc: 0
+  end
+
+  def key=(key)
+    raise ArgumentError.new "key length too short: wanted #{key_len}, got #{key.bytesize}" if key.bytesize < key_len
+    cipherinit key: key
+    key
+  end
+
+  def iv=(iv)
+    raise ArgumentError.new "iv length too short: wanted #{iv_len}, got #{iv.bytesize}" if iv.bytesize < iv_len
+    cipherinit iv: iv
+    iv
+  end
+
+  def random_key
+    key = SecureRandom.random_bytes key_len
+    self.key = key
+  end
+
+  def random_iv
+    iv = SecureRandom.random_bytes iv_len
+    self.iv = iv
+  end
+
+  def reset
+    cipherinit
+  end
+
+  def update(data : (String | Slice))
+    slice = case data
+      when String
+        data.to_slice
+      else
+        data
+    end
+
+    buffer_length = slice.size + block_size
+    buffer = Slice(UInt8).new(buffer_length)
+    if LibCrypto.evp_cipherupdate(@ctx, buffer, pointerof(buffer_length), slice, slice.size) != 1
+      raise Error.new "EVP_CipherUpdate"
+    end
+
+    buffer[0, buffer_length]
+  end
```

Comment:
I don't know if that will work. In Digest the evp_digestupdate does not require a buffer to store the info in. The evp_cipherupdate does require a separate buffer. So it will only work if Cipher#<< stores the buffer in the instance